### DISCUSSION TITLE
Noob question about deployment

### DISCUSSION THREAD

**Discussion by machty** [RailsEventStore/rails_event_store#1685] [Q&A]
I'm most familiar with a classic Heroku setup where all server processes are declared in a Procfile (web, workers, potentially multiple workers instances dedicated to a particular Sidekiq queue), and I'm curious about RES.

I've read [these docs](https://railseventstore.org/docs/v2/subscribe/) on subscribing to events, but I'm confused as to what process(es) (web, workers, etc) are actually going to be subscribing to queues. So, a few questions:

1. What, if anything, am I adding to the Procfile? Will there be a declaration of a new type of process, one that's a dedicate RES client that subscribes to various queues?
2. What would/should the concurrency be for that client? Is it expected to have 2+ instances consuming the same queues, or does that kinda defeat the point (and cause needless locking contention)? Is it expected that only during server turnover that there'd be > 1 instance consuming from a queue
3. What/where is the actually bootstrapping code to spin up such a client?

**Comment by mostlyobvious** [comment]
The pub-sub mechanism in RES, as of current version, is in-process. That is you publish an event and in the same process (let's say in web in Procfile) it is dispatched to a handler. 

In case of a synchronous handler it is executed in the very same process too. When your [handler is asynchronous](https://railseventstore.org/docs/v2/subscribe/#async-handlers) on the other hand, the queueing happens via ActiveJob and the handler is executed depending on the underlying AJ driver. In case of Sidekiq that would be an external Sidekiq process (worker in Procfile).

If you have a working ActiveJob subsystem, nothing needs to be added. There's a possibility to skip ActiveJob subsystem — by having your own dispatcher/scheduler components and a client to consume scheduled events. 

Assuming the AJ subsystem is at-least-once (Sidekiq is), you need a way to ensure idempotency. 

Most likely the AJ subsystem will handle events without order guarantee, which [has to be addressed as well](https://blog.arkency.com/read-model-patterns-in-case-of-lack-of-order-guarantee/).
